src/mlpack/core/dists/discrete_distribution.cpp

```diff
@@ -44,67 +44,97 @@ arma::vec DiscreteDistribution::Random() const
-  probabilities.zeros();
+  // Deal with the difference between vector and martix
+  const arma::mat transObservations = observations.t();
 
-  // Add the probability of each observation.  The addition of 0.5 to the
-  // observation is to turn the default flooring operation of the size_t cast
-  // into a rounding operation.
-  for (size_t i = 0; i < observations.n_cols; i++)
+  // Make sure the observations have same dimension with the probabilities
+  if(transObservations.n_cols != probabilities.size())
   {
-    const size_t obs = size_t(observations(0, i) + 0.5);
-
-    // Ensure that the observation is within the bounds.
-    if (obs >= probabilities.n_elem)
+    Log::Debug << "the obversation must has the same dimension with the probabilities"
+        << "the observation's dimension is" << transObservations.n_cols << "but the dimension of "
+        << "probabilities is" << probabilities.size() << std::endl;
+  }
+  // Get the dimension size of the distribution
+  const size_t dimensions = probabilities.size();
+  // Iterate all the probabilities in each dimension
+  for (size_t i=0; i < dimensions; i++)
+  {
+    // Clear the old probabilities
+    probabilities[i].zeros();
+    for (size_t r=0; r < transObservations.n_rows; r++)
     {
-      Log::Debug << "DiscreteDistribution::Train(): observation " << i
-          << " (" << obs << ") is invalid; observation must be in [0, "
-          << probabilities.n_elem << "] for this distribution." << std::endl;
+      // Add the probability of each observation.  The addition of 0.5 to the
+      // observation is to turn the default flooring operation of the size_t cast
+      // into a rounding observation.
+      const size_t obs = size_t(transObservations(r, i) + 0.5);
+
+      // Ensure that the observation is within the bounds.
+      if (obs >= probabilities[i].n_elem)
+      {
+        Log::Debug << "DiscreteDistribution::Train(): observation " << i
+            << " (" << obs << ") is invalid; observation must be in [0, "
+            << probabilities[i].n_elem << "] for this distribution." << std::endl;       
+      }
+      probabilities[i][obs]++;
     }
 
-    probabilities(obs)++;
+    // Now normailze the distribution.
+    double sum = accu(probabilities[i]);
+    if (sum > 0)
+      probabilities[i] /= sum;
+    else
+      probabilities[i].fill(1.0 / probabilities[i].n_elem); // Force normalization.
   }
-
-  // Now normalize the distribution.
-  double sum = accu(probabilities);
-  if (sum > 0)
-    probabilities /= sum;
-  else
-    probabilities.fill(1 / probabilities.n_elem); // Force normalization.
 }
 
 /**
  * Estimate the probability distribution from the given observations when also
  * given probabilities that each observation is from this distribution.
  */
 void DiscreteDistribution::Train(const arma::mat& observations,
-                                    const arma::vec& probObs)
+                                    const arma::mat& probObs)
```

Comment:
Hi @rcurtin , I still haven't got the point here after I try it. For example, the test in here https://github.com/mlpack/mlpack/pull/830/files#diff-ad93795d91105ba0fcc6e9815d5e9ad0R106

The observations is one dimensional **("0 0 1 2")** and prob is **("0.25 0.25 0.5 1.0")**, and **i** as an indicate of dimension will keep staying in **0** which means one dimension.

When we have the line

> probabilities[i][obs] += probObs[i];


But here comes to the problem, the probObs[i] is always 0.25 after loop over, the `probabilities` will become (0.5[0, 0], 0.25[1], 0.25[2]).

Maybe you mean the `probabilities[i][obs] += probObs[obs];`? If so, what will happened if it's a multidimensional observation like 

>"0 1  1 2 2 2;"
>"0 0 0 1  1 2;"  
>"0 1  1 2  2 2;"